lesson_10/datasource/datasource.go

```diff
@@ -0,0 +1,127 @@
+package datasource
+
+import (
+	"context"
+	"errors"
+	"log"
+
+	_ "github.com/lib/pq"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+	"go.mongodb.org/mongo-driver/mongo/readpref"
+)
+
+const (
+	database = "tasker"
+	collGr   = "groups"
+	collCont = "contacts"
+)
+
+// create new connection in database
+func NewDB(ctx context.Context) *mongo.Client {
+	conn_url := "mongodb://root:rootpassword@localhost:27017/?authSource=admin&readPreference=primary&appname=mongodb-vscode%200.6.10&ssl=false"
+
+	_ = "mongodb+srv://root:rootpassword@localhost:27017/corp?retryWrites=true&w=majority"
+	client, err := mongo.NewClient(options.Client().ApplyURI(conn_url))
+	if err != nil {
+		log.Fatal(err)
+	}
+	err = client.Connect(ctx)
+	if err != nil {
+		log.Fatal(err)
+	}
+	err = client.Ping(ctx, readpref.Primary())
+	if err != nil {
+		log.Fatal(err)
+	}
+	log.Println("Connection to mongo is ready")
+
+	return client
+}
+
+type DS struct {
+	DB *mongo.Client
+}
+
+func NewDS(conn *mongo.Client) DS {
+	return DS{
+		DB: conn,
+	}
+}
+
+// add contacts in database
+func (ds *DS) AddContact(ctx context.Context, name, number string) error {
+	collection := ds.DB.Database(database).Collection(collCont)
+	t := Contact{}
+	if err := ds.DB.Database(database).Collection(collCont).FindOne(ctx, bson.M{"name": name, "number": number}).Decode(&t); err == nil {
+		err1 := errors.New("Duplicates forbidden")
+		return err1
+	}
+	var c Contact
```

Comment:
I prefer use literals in cases like that, 
for example: c := Contact { Name: name, Number: number}
it's up to you, it's not necessary to fix.